File: src/main/java/com/skysoft/mixin/AbstractContainerScreenAccessor.java
package com.skysoft.mixin;

import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen;
import net.minecraft.client.gui.GuiGraphicsExtractor;
import net.minecraft.resources.Identifier;
import net.minecraft.world.inventory.ContainerInput;
import net.minecraft.world.inventory.Slot;
import org.spongepowered.asm.mixin.Mixin;
import org.spongepowered.asm.mixin.gen.Accessor;
import org.spongepowered.asm.mixin.gen.Invoker;

@Mixin(AbstractContainerScreen.class)
public interface AbstractContainerScreenAccessor {
    @Accessor("leftPos")
    int skysoft$getLeftPos();

    @Accessor("topPos")
    int skysoft$getTopPos();

    @Accessor("imageWidth")
    int skysoft$getImageWidth();

    @Accessor("imageHeight")
    int skysoft$getImageHeight();

    @Accessor("skipNextRelease")
    void skysoft$setSkipNextRelease(boolean skipNextRelease);

    @Accessor("hoveredSlot")
    Slot skysoft$getHoveredSlot();

    @Accessor("SLOT_HIGHLIGHT_BACK_SPRITE")
    static Identifier skysoft$getSlotHighlightBackSprite() {
        throw new AssertionError();
    }

    @Accessor("SLOT_HIGHLIGHT_FRONT_SPRITE")
    static Identifier skysoft$getSlotHighlightFrontSprite() {
        throw new AssertionError();
    }

    @Invoker("extractSlot")
    void skysoft$extractSlot(GuiGraphicsExtractor context, Slot slot, int mouseX, int mouseY);

    @Invoker("slotClicked")
    void skysoft$slotClicked(Slot slot, int slotId, int button, ContainerInput action);
}

File: src/main/kotlin/com/skysoft/mixin/AbstractContainerScreenAccessor.kt
package com.skysoft.mixin

import net.minecraft.client.gui.GuiGraphicsExtractor
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen
import net.minecraft.world.inventory.ContainerInput
import net.minecraft.world.inventory.Slot
import org.spongepowered.asm.mixin.Mixin
import org.spongepowered.asm.mixin.Mutable
import org.spongepowered.asm.mixin.gen.Accessor
import org.spongepowered.asm.mixin.gen.Invoker

@Mixin(AbstractContainerScreen::class)
interface AbstractContainerScreenAccessor {
    @Accessor("leftPos")
    fun skysoftGetLeftPos(): Int

    @Accessor("topPos")
    fun skysoftGetTopPos(): Int

    @Accessor("topPos")
    fun skysoftSetTopPos(topPos: Int)

    @Accessor("imageWidth")
    fun skysoftGetImageWidth(): Int

    @Accessor("imageHeight")
    fun skysoftGetImageHeight(): Int

    @Mutable
    @Accessor("imageHeight")
    fun skysoftSetImageHeight(imageHeight: Int)

    @Accessor("inventoryLabelY")
    fun skysoftSetInventoryLabelY(inventoryLabelY: Int)

    @Accessor("skipNextRelease")
    fun skysoftSetSkipNextRelease(skipNextRelease: Boolean)

    @Accessor("hoveredSlot")
    fun skysoftGetHoveredSlot(): Slot?

    @Invoker("extractSlot")
    fun skysoftExtractSlot(context: GuiGraphicsExtractor, slot: Slot, mouseX: Int, mouseY: Int)

    @Invoker("slotClicked")
    fun skysoftSlotClicked(slot: Slot?, slotId: Int, button: Int, action: ContainerInput)
}
